File: extension/m3u8.js
/** @typedef {{ url: string, bandwidth: number, resolution: string }} Variant */

function shouldCaptureVideoRequest(url) {
  if (!url) return false;
  if (/\/api\/playlist\/master\//i.test(url)) return true;
  if (/\/api\/playlist\/media\//i.test(url)) return true;
  if (/gceuproxy\.com/i.test(url) && /playlist/i.test(url)) return true;
  if (/\.m3u8(\?|$)/i.test(url)) return true;
  if (/\.(mp4|webm|m4v)(\?|$)/i.test(url)) return true;
  return false;
}

function isMasterPlaylistUrl(url) {
  return /\/api\/playlist\/master\//i.test(url) || /master\.m3u8/i.test(url);
}

function resolvePlaylistUrl(line, baseUrl) {
  const t = line.trim();
  if (!t || t.startsWith("#")) return null;
  if (/^https?:\/\//i.test(t)) return t;
  try {
    return new URL(t, baseUrl).href;
  } catch {
    return null;
  }
}

/** Parse #EXTM3U master or media playlist — all segment / variant URLs. */
function parseM3U8Variants(text, baseUrl) {
  /** @type {Variant[]} */
  const variants = [];
  const lines = text.split(/\r?\n/);
  let pendingBw = 0;
  let pendingRes = "";

  for (const raw of lines) {
    const line = raw.trim();
    if (line.startsWith("#EXT-X-STREAM-INF")) {
      const bw = line.match(/BANDWIDTH=(\d+)/i);
      const res = line.match(/RESOLUTION=(\d+x\d+)/i);
      pendingBw = bw ? parseInt(bw[1], 10) : 0;
      pendingRes = res ? res[1] : "";
      continue;
    }
    const url = resolvePlaylistUrl(line, baseUrl);
    if (!url) continue;
    variants.push({
      url,
      bandwidth: pendingBw,
      resolution: pendingRes,
    });
    pendingBw = 0;
    pendingRes = "";
  }
  return variants;
}

/** One logical stream (same media ids, different resolution). */
function streamGroupKey(url) {
  try {
    const u = new URL(url);
    const m = u.pathname.match(/\/media\/([^/]+)\/([^/]+)\//i);
    if (m) return `${u.host}|${m[1]}|${m[2]}`;
    const master = u.pathname.match(/\/master\/([^/]+)\/([^/]+)/i);
    if (master) return `${u.host}|master|${master[1]}|${master[2]}`;
    return `${u.host}|${u.pathname.split("?")[0]}`;
  } catch {
    return url;
  }
}

function resolutionFromMediaPath(url) {
  const m = String(url).match(/\/(\d{3,4})(?:\?|$|\/)/);
  return m ? parseInt(m[1], 10) : 0;
}

/** One master or one media stream per video on the lesson page. */
function selectVideoInputs(urls) {
  const masters = [];
  const seenMaster = new Set();
  for (const u of urls) {
    if (!isMasterPlaylistUrl(u)) continue;
    const k = streamGroupKey(u);
    if (seenMaster.has(k)) continue;
    seenMaster.add(k);
    masters.push(u);
  }
  if (masters.length) return masters;

  const mediaBest = new Map();
  for (const u of urls) {
    if (!/\/api\/playlist\/media\//i.test(u)) continue;
    const k = streamGroupKey(u);
    const score = resolutionFromMediaPath(u);
    const prev = mediaBest.get(k);
    if (!prev || score > prev.score) mediaBest.set(k, { url: u, score });
  }
  return [...mediaBest.values()].map((x) => x.url);
}

/** Keep highest BANDWIDTH per stream; if tie, prefer larger resolution. */
function pickBestVariants(variants) {
  const best = new Map();
  for (const v of variants) {
    const key = streamGroupKey(v.url);
    const prev = best.get(key);
    if (!prev || variantScore(v) > variantScore(prev)) {
      best.set(key, v);
    }
  }
  return [...best.values()];
}

function variantScore(v) {
  let score = v.bandwidth || 0;
  const m = (v.resolution || "").match(/(\d+)x(\d+)/);
  if (m) score += parseInt(m[1], 10) * parseInt(m[2], 10);
  return score;
}

async function expandMasterPlaylists(urls, fetchHeaders) {
  const inputs = selectVideoInputs(urls);
  const all = [];
  const seen = new Set();

  const addVariant = (v) => {
    if (!v?.url || seen.has(v.url)) return;
    seen.add(v.url);
    all.push(v);
  };

  for (const url of inputs) {
    if (!isMasterPlaylistUrl(url)) {
      addVariant({ url, bandwidth: 0, resolution: "" });
      continue;
    }
    try {
      const r = await fetch(url, {
        credentials: "omit",
        headers: fetchHeaders,
      });
      if (!r.ok) {
        addVariant({ url, bandwidth: 0, resolution: "" });
        continue;
      }
      const text = await r.text();
      if (!text.includes("#EXTM3U")) {
        addVariant({ url, bandwidth: 0, resolution: "" });
        continue;
      }
      for (const v of parseM3U8Variants(text, url)) {
        addVariant(v);
      }
    } catch {
      addVariant({ url, bandwidth: 0, resolution: "" });
    }
  }

  const picked = pickBestVariants(all);
  if (!picked.length) {
    for (const url of inputs) {
      addVariant({ url, bandwidth: 0, resolution: "" });
    }
  }
  return picked.length ? picked : all;
}

globalThis.gcM3U8 = {
  shouldCaptureVideoRequest,
  isMasterPlaylistUrl,
  parseM3U8Variants,
  selectVideoInputs,
  expandMasterPlaylists,
  pickBestVariants,
};

File: desktop/internal/m3u8/m3u8.go
package m3u8

import (
	"bufio"
	"net/http"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

var (
	reBandwidth  = regexp.MustCompile(`(?i)BANDWIDTH=(\d+)`)
	reResolution = regexp.MustCompile(`(?i)RESOLUTION=(\d+x\d+)`)
	reMaster     = regexp.MustCompile(`(?i)/api/playlist/master/`)
)

type Variant struct {
	URL        string
	Bandwidth  int
	Resolution string
}

func IsMasterPlaylistURL(u string) bool {
	return reMaster.MatchString(u)
}

func ParseVariants(text, baseURL string) []Variant {
	var out []Variant
	sc := bufio.NewScanner(strings.NewReader(text))
	var bw int
	var res string
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "#EXT-X-STREAM-INF") {
			if m := reBandwidth.FindStringSubmatch(line); m != nil {
				bw, _ = strconv.Atoi(m[1])
			}
			if m := reResolution.FindStringSubmatch(line); m != nil {
				res = m[1]
			}
			continue
		}
		if strings.HasPrefix(line, "#") {
			continue
		}
		abs := resolveURL(line, baseURL)
		if abs == "" {
			continue
		}
		out = append(out, Variant{URL: abs, Bandwidth: bw, Resolution: res})
		bw = 0
		res = ""
	}
	return out
}

func resolveURL(ref, base string) string {
	ref = strings.TrimSpace(ref)
	if ref == "" || strings.HasPrefix(ref, "#") {
		return ""
	}
	if strings.HasPrefix(ref, "http://") || strings.HasPrefix(ref, "https://") {
		return ref
	}
	bu, err := url.Parse(base)
	if err != nil {
		return ref
	}
	ru, err := url.Parse(ref)
	if err != nil {
		return ""
	}
	return bu.ResolveReference(ru).String()
}

func streamKey(u string) string {
	pu, err := url.Parse(u)
	if err != nil {
		return u
	}
	path := pu.Path
	if i := strings.Index(path, "/media/"); i >= 0 {
		rest := path[i+len("/media/"):]
		parts := strings.Split(rest, "/")
		if len(parts) >= 2 {
			return pu.Host + "|" + parts[0] + "|" + parts[1]
		}
	}
	if i := strings.Index(path, "/master/"); i >= 0 {
		rest := path[i+len("/master/"):]
		parts := strings.Split(rest, "/")
		if len(parts) >= 2 {
			return pu.Host + "|master|" + parts[0] + "|" + parts[1]
		}
	}
	return pu.Host + "|" + path
}

func variantScore(v Variant) int {
	score := v.Bandwidth
	if v.Resolution != "" {
		parts := strings.Split(v.Resolution, "x")
		if len(parts) == 2 {
			w, _ := strconv.Atoi(parts[0])
			h, _ := strconv.Atoi(parts[1])
			score += w * h
		}
	}
	return score
}

func PickBest(variants []Variant) []Variant {
	best := make(map[string]Variant)
	for _, v := range variants {
		k := streamKey(v.URL)
		if prev, ok := best[k]; !ok || variantScore(v) > variantScore(prev) {
			best[k] = v
		}
	}
	out := make([]Variant, 0, len(best))
	for _, v := range best {
		out = append(out, v)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].URL < out[j].URL })
	return out
}

func resolutionFromMediaPath(u string) int {
	re := regexp.MustCompile(`/(\d{3,4})(?:\?|$|/)`)
	m := re.FindStringSubmatch(u)
	if m == nil {
		return 0
	}
	n, _ := strconv.Atoi(m[1])
	return n
}

// SelectVideoInputs keeps one master (or best media) per video stream.
func SelectVideoInputs(urls []string) []string {
	seenMaster := make(map[string]bool)
	var masters []string
	for _, u := range urls {
		u = strings.TrimSpace(u)
		if u == "" || !IsMasterPlaylistURL(u) {
			continue
		}
		k := streamKey(u)
		if seenMaster[k] {
			continue
		}
		seenMaster[k] = true
		masters = append(masters, u)
	}
	if len(masters) > 0 {
		return masters
	}
	bestMedia := make(map[string]struct {
		url   string
		score int
	})
	for _, u := range urls {
		u = strings.TrimSpace(u)
		if !strings.Contains(strings.ToLower(u), "/api/playlist/media/") {
			continue
		}
		k := streamKey(u)
		sc := resolutionFromMediaPath(u)
		if prev, ok := bestMedia[k]; !ok || sc > prev.score {
			bestMedia[k] = struct {
				url   string
				score int
			}{u, sc}
		}
	}
	out := make([]string, 0, len(bestMedia))
	for _, v := range bestMedia {
		out = append(out, v.url)
	}
	sort.Strings(out)
	return out
}

// ExpandMasters fetches master playlists and returns best media URL per stream.
func ExpandMasters(client *http.Client, headers http.Header, urls []string) []string {
	urls = SelectVideoInputs(urls)
	var all []Variant
	for _, u := range urls {
		u = strings.TrimSpace(u)
		if u == "" {
			continue
		}
		if !IsMasterPlaylistURL(u) {
			all = append(all, Variant{URL: u})
			continue
		}
		req, err := http.NewRequest(http.MethodGet, u, nil)
		if err != nil {
			all = append(all, Variant{URL: u})
			continue
		}
		for k, vv := range headers {
			for _, v := range vv {
				req.Header.Add(k, v)
			}
		}
		resp, err := client.Do(req)
		if err != nil {
			all = append(all, Variant{URL: u})
			continue
		}
		body := resp.Body
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			body.Close()
			all = append(all, Variant{URL: u})
			continue
		}
		var b strings.Builder
		sc := bufio.NewScanner(body)
		for sc.Scan() {
			b.WriteString(sc.Text())
			b.WriteByte('\n')
		}
		body.Close()
		text := b.String()
		if !strings.Contains(text, "#EXTM3U") {
			all = append(all, Variant{URL: u})
			continue
		}
		all = append(all, ParseVariants(text, u)...)
	}
	picked := PickBest(all)
	out := make([]string, len(picked))
	for i, v := range picked {
		out[i] = v.URL
	}
	return out
}
